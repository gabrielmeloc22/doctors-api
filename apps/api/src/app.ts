import bodyParser from "body-parser"
import cors from "cors"
import express, { type Express } from "express"
import morgan from "morgan"

export const createApp = (): Express => {
    const app = express()

    app.disable("x-powered-by")
        .use(morgan("dev"))
        .use(bodyParser.urlencoded({ extended: true }))
        .use(bodyParser.json())
        .use(cors())

    return app
}
