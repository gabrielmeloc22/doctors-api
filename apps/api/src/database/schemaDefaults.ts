import { randomUUID } from "crypto";
import * as t from "drizzle-orm/pg-core";

export const idField = {
    id: t.uuid().primaryKey().$defaultFn(randomUUID),
};

export const timestampField = {
    createdAt: t.timestamp().notNull().defaultNow(),
    updatedAt: t
        .timestamp()
        .notNull()
        .defaultNow()
        .$defaultFn(() => new Date()),
};
