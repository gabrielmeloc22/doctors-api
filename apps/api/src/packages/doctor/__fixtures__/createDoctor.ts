import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { doctorTable, IDoctor } from "../doctorTable";

export const createDoctor = async (
    args?: DeepPartial<IDoctor>
): Promise<IDoctor> => {
    const doctorCount = await db.$count(doctorTable);

    const [doctor] = await db
        .insert(doctorTable)
        .values({
            email: `doc${doctorCount}@mail.com`,
            firstName: "john",
            lastName: `doe ${doctorCount}`,
            username: `john#${doctorCount}`,
            ...args,
        })
        .returning();

    return doctor as IDoctor;
};
