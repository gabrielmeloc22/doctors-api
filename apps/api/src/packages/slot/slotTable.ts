import * as t from "drizzle-orm/pg-core";
import { idField, timestampField } from "../database/schemaDefaults";
import { doctorTable } from "../doctor/doctorTable";

export const slotTable = t.pgTable("slot", {
    doctorId: t.uuid().references(() => doctorTable.id),
    startTime: t.date().notNull(),
    endTime: t.date().notNull(),
    available: t.boolean(),
    ...timestampField,
    ...idField,
});
