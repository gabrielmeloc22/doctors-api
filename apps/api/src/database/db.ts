import { drizzle } from "drizzle-orm/node-postgres";
import * as booking from "../packages/booking/bookingTable";
import * as doctor from "../packages/doctor/doctorTable";
import * as slot from "../packages/slot/slotTable";

export const schema = {
    ...slot,
    ...doctor,
    ...booking,
};

export const db = drizzle({
    connection: process.env.DATABASE_URL,
    casing: "camelCase",
    schema,
});
