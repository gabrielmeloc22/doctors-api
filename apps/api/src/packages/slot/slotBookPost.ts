import { getDay, getHours } from "date-fns";
import { and, eq } from "drizzle-orm";
import { RequestHandler } from "express";
import { z } from "zod";
import { db } from "../../database/db";
import { Result } from "../../utils/result";
import { validateRequest } from "../../utils/validateRequest";
import { bookingTable } from "../booking/bookingTable";
import { SLOT_REPEAT_TYPE_ENUM, slotTable } from "../slot/slotTable";
import { getTimeSlot } from "./lib/getTimeSlot";
import { getTimeSlots } from "./lib/getTimeSlots";

const slotBookPostParams = z.object({
    id: z.string().uuid(),
});

type SlotBookPostParams = z.infer<typeof slotBookPostParams>;

const slotBookPostBodySchema = z.object({
    reason: z.string(),
    patient_id: z.string(),
    start_time: z.string().datetime(),
});

type SlotBookPostBody = z.infer<typeof slotBookPostBodySchema>;

type SlotBookPostResult = Result<{
    booking: {
        id: string;
        slot_id: string;
        patient_id: string;
        reason: string;
        time: string;
    };
}>;

const slotBookPostHandler: RequestHandler<
    SlotBookPostParams,
    SlotBookPostResult,
    SlotBookPostBody
> = async (req, res) => {
    const [slot] = await db
        .select()
        .from(slotTable)
        .where(eq(slotTable.id, req.params.id));

    if (!slot) {
        res.status(404);
        res.json({ success: false, error: "Slot not found" });
        return;
    }

    // TODO: create recurrence rule validator

    const selectedStart = new Date(req.body.start_time);

    if (selectedStart < slot.startTime) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    if (slot.repeatEnd && selectedStart > new Date(slot.repeatEnd)) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    if (
        slot.repeatType === SLOT_REPEAT_TYPE_ENUM.WEEKLY &&
        !slot.repeatWeekdays?.includes(getDay(selectedStart))
    ) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    if (getHours(selectedStart) > getHours(slot.endTime)) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    const timeSlots = getTimeSlots(slot);

    const selectedTimeSlot = getTimeSlot(slot, selectedStart);

    // check if booking time is within time slots of the day
    if (timeSlots[selectedTimeSlot] === undefined) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    const [existingBooking] = await db
        .select()
        .from(bookingTable)
        .where(
            and(
                eq(bookingTable.time, selectedStart),
                eq(bookingTable.slotId, slot.id)
            )
        );

    if (existingBooking) {
        res.status(400);
        res.json({ success: false, error: "Booking time already taken" });
    }

    const [booking] = await db
        .insert(bookingTable)
        .values({
            patientId: req.body.patient_id,
            reason: req.body.reason,
            time: selectedStart,
            slotId: slot.id,
        })
        .returning();

    if (!booking) {
        res.status(400);
        res.json({ success: false, error: "Error while booking time" });

        return;
    }

    res.status(201);
    res.json({
        success: true,
        booking: {
            id: booking.id,
            patient_id: booking.patientId,
            reason: booking.reason,
            slot_id: booking.slotId,
            time: booking.time.toISOString(),
        },
    });
};

export const slotBookPost: RequestHandler[] = [
    validateRequest({
        body: slotBookPostBodySchema,
        params: slotBookPostParams,
    }),
    slotBookPostHandler as RequestHandler,
];
