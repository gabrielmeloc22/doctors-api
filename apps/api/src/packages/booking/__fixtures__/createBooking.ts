import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { createSlot } from "../../slot/__fixtures__/createSlot";
import { bookingTable, IBooking } from "../bookingTable";

export const createBooking = async (
    args?: DeepPartial<IBooking>
): Promise<IBooking> => {
    const slotId = args?.slotId ? args.slotId : (await createSlot()).id;

    const [booking] = await db
        .insert(bookingTable)
        .values({
            slotId: slotId,
            time: new Date(),
            patientId: "1",
            reason: "checkup",
            ...args,
        })
        .returning();

    return booking as IBooking;
};
