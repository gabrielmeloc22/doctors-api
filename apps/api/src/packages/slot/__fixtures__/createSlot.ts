import { add } from "date-fns";
import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { createDoctor } from "../../doctor/__fixtures__/createDoctor";
import { ISlot, SLOT_REPEAT_TYPE_ENUM, slotTable } from "../slotTable";

export const createSlot = async (args?: DeepPartial<ISlot>): Promise<ISlot> => {
    const doctor = await createDoctor();

    const [slot] = await db
        .insert(slotTable)
        .values({
            doctorId: doctor.id,
            duration: 30,
            startTime: new Date(),
            endTime: add(new Date(), { minutes: 30 }),
            repeatType: SLOT_REPEAT_TYPE_ENUM.SINGLE,
            ...args,
        })
        .returning();

    return slot as ISlot;
};
