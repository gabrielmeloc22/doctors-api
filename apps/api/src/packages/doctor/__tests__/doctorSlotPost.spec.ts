import request from "supertest";
import { expect, it } from "vitest";
import { createApp } from "../../../app";
import { db } from "../../../database/db";
import {
    defaultFrozenKeys,
    sanitizeTestObject,
} from "../../../testUtils/sanitizeTestObject";
import { SLOT_REPEAT_TYPE_ENUM, slotTable } from "../../slot/slotTable";
import { createDoctor } from "../__fixtures__/createDoctor";

it("should validate params and return 400", async () => {
    const app = createApp();

    const doctor = await createDoctor();

    const payload = {
        doctor_id: doctor.id,
        start_time: new Date("2025-01-01T08:00:00.000Z"),
        end_time: new Date("2025-01-01T08:30:00.000Z"),
        duration: 30,
        repeat: {
            type: SLOT_REPEAT_TYPE_ENUM.DAILY,
            end: new Date("2025-01-01T08:00:00.000Z"),
            days: [1, 3, 5], // monday, wednesday, friday
        },
    };

    const response = await request(app.listen())
        .post("/doctors/0a523035-fcca-4bbd-a7e6-3918e4689/slots")
        .send(payload);

    const body = response.body as Record<string, unknown>;

    expect(response.status).toBe(400);

    expect(body).toMatchSnapshot();
});

it("should validate body and return 400", async () => {
    const app = createApp();

    const doctor = await createDoctor();

    const payload = {
        start_time: "",
        end_time: "",
        duration: 50,
        repeat: {
            type: "type",
            end: "",
            days: [1, 3, 5, 8],
        },
    };

    const response = await request(app.listen())
        .post(`/doctors/${doctor.id}/slots`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    expect(response.status).toBe(400);

    expect(body).toMatchSnapshot();
});

it("should post, create a new slot and return 201", async () => {
    const app = createApp();

    const doctor = await createDoctor();

    const payload = {
        doctor_id: doctor.id,
        start_time: new Date("2025-01-01T08:00:00.000Z"),
        end_time: new Date("2025-01-01T08:30:00.000Z"),
        duration: 30,
        repeat: {
            type: SLOT_REPEAT_TYPE_ENUM.DAILY,
            end: new Date("2025-01-01T08:00:00.000Z"),
            days: [1, 3, 5], // monday, wednesday, friday
        },
    };

    const response = await request(app.listen())
        .post(`/doctors/${doctor.id}/slots`)
        .send(payload);

    const slots = await db.select().from(slotTable);
    const [slot] = slots;

    const slotResult = {
        doctor_id: slot?.doctorId,
        duration: slot?.duration,
        start_time: slot?.startTime.toISOString(),
        end_time: slot?.endTime.toISOString(),
        repeat: {
            type: slot?.repeatType,
            days: slot?.repeatWeekdays,
            end: slot?.endTime.toISOString(),
        },
    };
    const body = response.body as Record<string, unknown>;

    expect(response.status).toBe(201);
    expect(slots).toHaveLength(1);

    expect(body).toMatchObject({
        success: true,
        slot: slotResult,
    });

    expect(
        sanitizeTestObject({
            obj: body,
            frozenKeys: [...defaultFrozenKeys, "doctor_id"],
        })
    ).toMatchSnapshot();
});

it("should post with slot start time greater than end time and return 400", async () => {
    const app = createApp();

    const doctor = await createDoctor();

    const payload = {
        doctor_id: doctor.id,
        start_time: new Date("2025-01-01T08:30:00.000Z"),
        end_time: new Date("2025-01-01T08:00:00.000Z"),
        duration: 30,
        repeat: {
            type: SLOT_REPEAT_TYPE_ENUM.DAILY,
            end: new Date("2025-01-01T08:00:00.000Z"),
            days: [1, 3, 5], // monday, wednesday, friday
        },
    };

    const response = await request(app.listen())
        .post(`/doctors/${doctor.id}/slots`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    expect(response.status).toBe(400);

    expect(body).toMatchObject({
        success: false,
        error: "Start time must be before end time",
    });
});

it("should post with start time and end time not divisible by duration and return 400", async () => {
    const app = createApp();

    const doctor = await createDoctor();

    const payload = {
        doctor_id: doctor.id,
        start_time: new Date("2025-01-01T08:00:00.000Z"),
        end_time: new Date("2025-01-01T08:40:00.000Z"),
        duration: 30,
        repeat: {
            type: SLOT_REPEAT_TYPE_ENUM.DAILY,
            end: new Date("2025-01-01T08:00:00.000Z"),
            days: [1, 3, 5], // monday, wednesday, friday
        },
    };

    const response = await request(app.listen())
        .post(`/doctors/${doctor.id}/slots`)
        .send(payload);

    const body = response.body as Record<string, unknown>;

    expect(response.status).toBe(400);

    expect(body).toMatchObject({
        success: false,
        error: "Invalid start and end time. Difference between start and end time must be an integer multiple of duration",
    });
});
