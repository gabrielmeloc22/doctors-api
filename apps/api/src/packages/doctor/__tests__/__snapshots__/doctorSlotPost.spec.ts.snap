// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should post, create a new slot and return 201 1`] = `
{
  "slot": {
    "doctor_id": "FROZEN-DOCTOR_ID",
    "duration": 30,
    "end_time": "2025-01-01T08:30:00.000Z",
    "id": "FROZEN-ID",
    "repeat": {
      "days": null,
      "end": "2025-01-01T08:30:00.000Z",
      "type": "DAILY",
    },
    "start_time": "2025-01-01T08:00:00.000Z",
  },
  "success": true,
}
`;

exports[`should validate body and return 400 1`] = `
{
  "error": [
    {
      "message": "Invalid datetime",
      "path": [
        "start_time",
      ],
    },
    {
      "message": "Invalid datetime",
      "path": [
        "end_time",
      ],
    },
    {
      "message": "Invalid input",
      "path": [
        "duration",
      ],
    },
    {
      "message": "Invalid input",
      "path": [
        "repeat",
      ],
    },
  ],
  "success": false,
}
`;

exports[`should validate params and return 400 1`] = `
{
  "error": [
    {
      "message": "Invalid uuid",
      "path": [
        "id",
      ],
    },
  ],
  "success": false,
}
`;
